File: apps/admin-web/src/utils/embedUrlBuilder.ts
/**
 * Extracts the 11-character video ID from various YouTube URL formats.
 */
export function extractYouTubeId(url: string): string | null {
  const regExp = /^.*(youtu.be\/|v\/|u\/\w\/|embed\/|watch\?v=|\&v=|shorts\/)([^#\&\?]*).*/i;
  const match = url.match(regExp);
  if (match && match[2] && match[2].length === 11) {
    return match[2];
  }
  return null;
}

/**
 * Extracts the numeric video ID from TikTok URL formats.
 */
export function extractTikTokVideoId(url: string): string | null {
  const regExp = /\/video\/(\d+)/i;
  const match = url.match(regExp);
  if (match && match[1]) {
    return match[1];
  }
  const regExpV = /\/v\/(\d+)/i;
  const matchV = url.match(regExpV);
  if (matchV && matchV[1]) {
    return matchV[1];
  }
  return null;
}

/**
 * Converts a stored externalUrl + platform into a proper embeddable URL.
 */
export function getEmbedUrl(externalUrl: string, platform: string): string | null {
  const cleanUrl = externalUrl.trim();
  const upperPlatform = platform.toUpperCase();

  if (upperPlatform === 'YOUTUBE') {
    const videoId = extractYouTubeId(cleanUrl);
    if (videoId) {
      return `https://www.youtube.com/embed/${videoId}?autoplay=0`;
    }
  } else if (upperPlatform === 'TIKTOK') {
    const videoId = extractTikTokVideoId(cleanUrl);
    if (videoId) {
      return `https://www.tiktok.com/embed/v2/${videoId}`;
    }
  } else if (upperPlatform === 'FACEBOOK') {
    return `https://www.facebook.com/plugins/video.php?href=${encodeURIComponent(cleanUrl)}&show_text=0&autoplay=false`;
  }

  return null;
}

File: apps/mobile/lib/core/utils/embed_url_builder.dart
class EmbedUrlBuilder {
  /// Converts a standard video URL and its platform into an embeddable URL.
  static String? getEmbedUrl(String externalUrl, String platform) {
    final cleanUrl = externalUrl.trim();
    final upperPlatform = platform.toUpperCase();

    if (upperPlatform == 'YOUTUBE') {
      final videoId = extractYouTubeId(cleanUrl);
      if (videoId != null) {
        return 'https://www.youtube.com/embed/$videoId?autoplay=1';
      }
    } else if (upperPlatform == 'TIKTOK') {
      final videoId = extractTikTokVideoId(cleanUrl);
      if (videoId != null) {
        return 'https://www.tiktok.com/embed/v2/$videoId';
      }
      return cleanUrl;
    } else if (upperPlatform == 'FACEBOOK') {
      // Facebook's Embedded Video Player Plugin URL is robust and handles redirects/watches.
      return 'https://www.facebook.com/plugins/video.php?href=${Uri.encodeComponent(cleanUrl)}&show_text=0&autoplay=true';
    }

    return cleanUrl;
  }

  /// Extracts the 11-character video ID from various YouTube URL formats.
  static String? extractYouTubeId(String url) {
    final regExp = RegExp(
      r'^.*(youtu.be\/|v\/|u\/\w\/|embed\/|watch\?v=|\&v=|shorts\/)([^#\&\?]*).*',
      caseSensitive: false,
    );
    final match = regExp.firstMatch(url);
    if (match != null && match.groupCount >= 2) {
      final id = match.group(2);
      if (id != null && id.length == 11) {
        return id;
      }
    }
    return null;
  }

  /// Extracts the numeric video ID from TikTok URL formats.
  static String? extractTikTokVideoId(String url) {
    final regExp = RegExp(r'\/video\/(\d+)', caseSensitive: false);
    final match = regExp.firstMatch(url);
    if (match != null && match.groupCount >= 1) {
      return match.group(1);
    }
    final regExpV = RegExp(r'\/v\/(\d+)', caseSensitive: false);
    final matchV = regExpV.firstMatch(url);
    if (matchV != null && matchV.groupCount >= 1) {
      return matchV.group(1);
    }
    return null;
  }
}
